tools/autograd/gen_variable_type.py

```diff
@@ -117,21 +117,28 @@
     'set', 'resize'
 }
 
-# NOTE [ Invariant: TensorImpl and Storage Pointer Equality ]
+# NOTE [ TensorImpl and Storage Pointer Sanity Checks ]
 #
-# When a function modifies its input tensors (via inplace or out-variants),
-# it should never change the the input tensors' underlying c10::TensorImpl pointers
-# or c10::Storage pointers.
+# We check the following properties:
+#   1) A function should never change the input tensors' underlying c10::TensorImpl
+#      pointers or c10::Storage pointers, even if it modifies its input tensors (via
+#      inplace or out-variants)
+# If the function does not modify its arguments, we also check the following properties
+# pertaining to its output:
+#   2) Its TensorImpl has use_count of 1
+#   3) If the function is a view function, it has the same StorageImpl as that of
+#      the input it is aliased with. Otherwise, its StorageImpl has use_count of 1
```

Comment:
There's a situation in functorch where a Tensor is not being aliased but its StorageImpl has use_count of > 1. I'm not sure if that's a problem or not, but right now the functorch tests are failing and I'm trying to understand what this check is. @soulitzer what is the reason why we check for use_count = 1 here?

Reply:
It's to make sure that operations that return aliases are properly marked as view ops in `gen_inplace_or_view_type` and have their view info set properly.